프로젝트 보완 기능 구현

seeyeon opened the pull request:
## 🌿 반영 브랜치
<!-- 작업 브랜치 -> 병합 대상 브랜치를 작성해주세요. -->  
`feature/project-FE2 -> develop`

## 🛠 작업 내용
<!-- 변경된 내용을 간단히 요약해주세요. -->  
- 프로젝트 등록 보완 ( 첨부파일, 설명부분에 직접 링크 `![]()`로 작성 가능하도록 구현, skill과 tool string이 아닌 드롭다운으로 구현)
- 프로젝트 전체 조회 보완 ( 내용이나 이름으로 검색)
- 프로젝트 단건 조회 보완 ( 전체 조회로 이동 버튼, 첨부파일 관련 내용 수정가능하도록 구현)
- 프로젝트 삭제 ( 휴지통 이동, 복구, 영구 삭제 구현)

## 🔗 관련 이슈
#81 

## 📂 추가 정보
<!-- 참고 자료나 관련 정보를 추가해주세요. -->  
1. 전체 조회 : https://i.postimg.cc/8C374R78/image.png
2. 단건 조회 : https://i.postimg.cc/0N6QzzcV/image.png
3. 수정 : https://i.postimg.cc/Y0tMz0zK/image.png
4. 삭제(휴지통) : https://i.postimg.cc/02pHqZvf/image.png 

## 💬 리뷰 요구사항 (선택)
<!-- 리뷰어에게 요청하고 싶은 내용이 있다면 작성해주세요. -->  
- 프론트 작업과 관련해서 BE 코드 수정이 있었습니다.

seeyeon committed:
bug : 첨부파일 관련 수정

본문: 추가 수정이 필요함

seeyeon committed:
refactor : thumbnailPath에 @JsonProperty추가

seeyeon committed:
bug : 첨부파일 관련 내용 수정

본문: 추가 수정이 필요한 부분

seeyeon committed:
refactor : FE 코드 임시 저장

seeyeon committed:
refactor : file 저장 경로 @Value 사용

seeyeon committed:
fix : 오류 해결을 위한 @Slf4j 사용 및 내용 일부 수정

seeyeon committed:
feat : 썸네일 관련 처리

seeyeon committed:
feat : 썸네일 관련 처리

seeyeon committed:
feat : 스킬과 도구 id -> 이름 가져오는 메서드 추가

seeyeon committed:
test : 썸네일 부분 추가해 등록Test 진행

seeyeon committed:
fix : [FE] 드롭다운 높이 조정

seeyeon committed:
fix : [FE] 라이브러리의 플러그인 설치

본문: 프로젝트 등록 시 설명에 이미지추가를 위해 설치함

seeyeon committed:
fix : [FE] 라이브러리의 플러그인 설치

본문: 프로젝트 등록 시 설명에 이미지추가를 위해 설치함

seeyeon committed:
fix : [FE]  프로젝트 수정

seeyeon committed:
fix : [FE] 전체 조회

seeyeon committed:
fix : [FE] 단건 조회

seeyeon committed:
fix : [FE] 단건 조회 로그

seeyeon committed:
fix : [FE] 프로젝트 등록

seeyeon committed:
fix : [FE] 프로젝트 등록,조회,수정,삭제 로직

seeyeon committed:
feat : [FE] 휴지통 관련 파일 추가 및 기본 프로젝트 이미지 추가

sangxxjin commented on backend/src/main/java/com/pofo/backend/domain/mapper/ProjectMapper.java:
```diff
@@ -8,6 +8,7 @@
 @Mapper(componentModel = "spring")
 public interface ProjectMapper {
     @Mapping(source = "id", target = "projectId")
+    @Mapping(source = "thumbnailPath", target = "thumbnailPath")
```
이거 삭제해도 자동으로 매핑되지 않나요? 단순 칼럼이 추가된거라 다른 칼럼처럼 명시 안해도 projectMapperImpl 파일이 변경되면서 될 것 같아요!

seeyeon replied:
동일한 필드명이라면 MapStruct가 자동 매핑하므로 삭제했습니다.

sangxxjin commented on backend/src/main/java/com/pofo/backend/domain/project/service/FileService.java:
```diff
@@ -51,7 +51,10 @@ public void deleteFile(String filePath) {
 
-        File file = new File(filePath);
+        // ✅ 저장된 파일의 물리적 경로를 가져와 삭제
+        String fileName = filePath.replace("http://localhost:8080/uploads/", "");
```
경로 env파일 활용하면 좋을 것 같네요! 그러면 경로 수정을 해야할때 41번줄, 55번줄 더 있다면 많이 수정해야하는데 env파일 활용하면 한줄만 수정하면 되는 이점이 있을 것 같습니다.

sangxxjin commented on backend/src/main/resources/application.yml:
```diff
@@ -85,39 +85,33 @@ spring:
-    properties:
-      mail.smtp.auth: true
-      mail.smtp.starttls.enable: true
+mail:
+  host: ${SPRING_MAIL_HOST}
+  port: ${SPRING_MAIL_PORT}
+  username: ${SPRING_MAIL_USERNAME}
+  password: ${SPRING_MAIL_PASSWORD}
+  properties:
+    mail.smtp.auth: true
+    mail.smtp.starttls.enable: true
```
yml파일은 앞의 공백도 중요한데 이렇게 작성이 되어버리면 "spring - mail - host" 이게 되는게 아니라 "mail - host"이렇게 적용이 돼서 에러가 발생할 것 같네요.

seeyeon replied:
✔수정완료

sangxxjin commented on backend/src/main/java/com/pofo/backend/domain/project/service/ProjectService.java:
프로젝트를 찾고 user를 통해 권한을 비교하는 로직들이 있는데 이런 부분들은 프로젝트를 찾을때 projectId와 user를 이용해서 한번에 찾으면 더 좋을 것 같아요!

seeyeon replied:
예외 처리가 더 명확해져서(notFound와 forbidden을 명확히 구분 가능하기 때문) 현재 코드를 유지하는게 더 좋을 것 같습니다.

sangxxjin commented on backend/src/main/resources/application.yml:
```diff
@@ -85,39 +85,33 @@ spring:
 
+  web:
+    resources:
+      static-locations: file:///C:/DevCourse_Project/NBE3-4-2-11-Team11/backend/uploads/
```
env파일 활용해야 할 것 같네요! 각자 로컬에서의 환경이 다릅니다!

seeyeon replied:
✔수정 완료

sangxxjin reviewed (commented):
이미지 처리 고생하셨습니다.  😊
백엔드 코드만 봤습니다. 리뷰 간단하게 남겨드렸습니다! 
개발이 완료되었으면 실제로 필요하지 않고 테스트할때 사용했던 주석, 로그는 제거해주는게 좋을 것 같아요!
수정할 부분만 수정하고 사소한 부분은 마이그레이션 단계에서 진행해주셔도 좋을 것 같아요. 😀

seeyeon committed:
fix : file 절대 경로로 수정 및 mail 경로 띄어쓰기 수정 완료

seeyeon committed:
fix : MapStruct가 자동 매핑하므로 삭제

seeyeon committed:
refactor : 설명 text로 수정해 긴 내용 저장

seeyeon committed:
refactor : refactor: JPQL 사용하여 TEXT 컬럼 upper() 문제 해결

본문: JPA의 TEXT(@Lob) 컬럼에서 upper() 함수가 동작하지 않는 문제 발생해 JPQL을 사용하여 CAST(description AS string) 방식으로 수정.

seeyeon committed:
refactor: 카드에서 설명이 긴 경우 1~2줄까지만 표시

seeyeon committed:
Merge branch 'develop' into feature/project-FE2
